File: tests/src/joda-time/joda-time/2.9.9/src/test/java/joda_time/joda_time/LocalDateTimeInnerPropertyTest.java
/*
 * Copyright and related rights waived via CC0
 *
 * You should have received a copy of the CC0 legalcode along with this
 * work. If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
 */
package joda_time.joda_time;

import static org.assertj.core.api.Assertions.assertThat;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.io.Serializable;

import org.joda.time.LocalDateTime;
import org.junit.jupiter.api.Test;

public class LocalDateTimeInnerPropertyTest {

    @Test
    void serializesAndDeserializesMinuteOfHourProperty() throws Exception {
        LocalDateTime.Property original = new LocalDateTime(2020, 2, 29, 12, 34, 56, 789).minuteOfHour();

        LocalDateTime.Property restored = deserialize(serialize(original));

        assertThat(restored.get()).isEqualTo(original.get());
        assertThat(restored.getFieldType()).isEqualTo(original.getFieldType());
        assertThat(restored.getLocalDateTime()).isEqualTo(original.getLocalDateTime());
    }

    private static byte[] serialize(Serializable value) throws IOException {
        ByteArrayOutputStream output = new ByteArrayOutputStream();
        try (ObjectOutputStream objectOutput = new ObjectOutputStream(output)) {
            objectOutput.writeObject(value);
        }
        return output.toByteArray();
    }

    private static LocalDateTime.Property deserialize(byte[] serialized) throws IOException, ClassNotFoundException {
        try (ObjectInputStream objectInput = new ObjectInputStream(new ByteArrayInputStream(serialized))) {
            return (LocalDateTime.Property) objectInput.readObject();
        }
    }
}

File: tests/src/org.playframework.anorm/anorm_3/2.8.1/src/test/scala/org_playframework_anorm/anorm_3/LocalDateTimeInnerPropertyTest.scala
/*
 * Copyright and related rights waived via CC0
 *
 * You should have received a copy of the CC0 legalcode along with this
 * work. If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
 */
package org_playframework_anorm.anorm_3

import java.io.ByteArrayInputStream
import java.io.ByteArrayOutputStream
import java.io.ObjectInputStream
import java.io.ObjectOutputStream

import org.joda.time.DateTimeFieldType
import org.joda.time.LocalDateTime
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test

class LocalDateTimeInnerPropertyTest {
  @Test
  def serializesAndDeserializesLocalDateTimeProperty(): Unit = {
    val originalDateTime: LocalDateTime = new LocalDateTime(2024, 5, 14, 9, 30, 15)
    val originalProperty: LocalDateTime.Property = originalDateTime.dayOfMonth()

    val serialized: Array[Byte] = serialize(originalProperty)
    val restoredProperty: LocalDateTime.Property = deserializeProperty(serialized)

    assertEquals(originalDateTime, restoredProperty.getLocalDateTime)
    assertEquals(DateTimeFieldType.dayOfMonth(), restoredProperty.getField.getType)
    assertEquals(14, restoredProperty.get)
  }

  private def serialize(property: LocalDateTime.Property): Array[Byte] = {
    val bytes: ByteArrayOutputStream = new ByteArrayOutputStream()
    val output: ObjectOutputStream = new ObjectOutputStream(bytes)
    try {
      output.writeObject(property)
    } finally {
      output.close()
    }
    bytes.toByteArray
  }

  private def deserializeProperty(serialized: Array[Byte]): LocalDateTime.Property = {
    val input: ObjectInputStream = new ObjectInputStream(new ByteArrayInputStream(serialized))
    try {
      input.readObject().asInstanceOf[LocalDateTime.Property]
    } finally {
      input.close()
    }
  }
}
